### PR TITLE
separate code if f is not defined

### LINKED ISSUE #20: Bug from function resolution
For MeasureTheory, I have a `@measure` macro that calls this `_measure` function:
```julia
function _measure(__module__, expr)
    ParameterizedMeasure = MeasureTheory.ParameterizedMeasure
    @capture $μ($(p...)) expr begin
        q = quote
            struct $μ{N,T} <: $ParameterizedMeasure{N}
                par :: NamedTuple{N,T}
            end
        end   

        q_kwstruct = KeywordCalls._kwstruct(__module__, expr)

        append!(q.args, q_kwstruct.args)
        
        if !isempty(p)
            # e.g. Normal(μ,σ) = Normal((μ=μ, σ=σ))
            pnames = QuoteNode.(p)
            push!(q.args, :($μ($(p...)) = $μ(NamedTuple{($(pnames...),)}(($(p...),)))))
        end
        
        return q
    end
end
```

When I call e.g. `@measure Normal` I get an `UndefVarError: Normal not defined`. The problem comes from the first few lines of `_kwcall`,
```julia
function _kwcall(__module__, ex)
    @assert Meta.isexpr(ex, :call)
    f = ex.args[1]
    args, defaults = _parse_args(ex.args[2:end])
    f_raw = getproperty(__module__, f)                # <---- bad stuff here
    ...
```

Here `f` is the symbol representing a function (or constructor), and `f_raw` is supposed to be the actual function. But in this case, the function (or constructor) doesn't yet exist!

### LINKED ISSUE #22: `@kwstruct` builds too-general methods
`@kwstruct` currently delegates some work to `_kwcall`, leading to methods like
```julia
_call_in_default_order(::UnionAll, nt::NamedTuple{(:μ, :σ), T} where T<:Tuple)
```

This happens because `_kwcall` includes this:
```julia
q = quote
    KeywordCalls._call_in_default_order(::typeof($f), nt::NamedTuple{($(sorted_args...),)}) = $f(NamedTuple{($(args...),)}(nt))
end
```

If `f` is a constructor, we end up with a `UnionAll`.

To handle this properly, we should have a new function, maybe... I don't know, `instance_type`? Let's go with that for now. So it should be

```julia
instance_type(f::Function) = typeof(f)
instance_type(f::UnionAll) = Type{f}
```

Not tested yet, but I think that's what we want.

### PR DESCRIPTION
**Pull request by cscherrer**

### CODE REVIEW

**Comment by cscherrer** [comment]
Hi @mileslucas @icweaver, I'm debugging some things that were breaking MeasureTheory. I think this is an improvement, can you check it with your package before I merge?

Code review feedback also welcome :)

**Comment by icweaver** [comment]
Nice! I am getting the same benchmark results on my end 👍🏾

**Comment by cscherrer** [comment]
Thanks @icweaver! Looks like it's just shy of 100% coverage. I'll update that before merging, that will give a little more time in case Miles has any comments

File: src/KeywordCalls.jl [hunk 1]
```diff
@@ -37,38 +38,45 @@ end
...
-    f = ex.args[1]
+    f_sym = ex.args[1]
+    f_esc = esc(f_sym)
     args, defaults = _parse_args(ex.args[2:end])
-    f_raw = getproperty(__module__, f)
-    f, args, sorted_args = esc(f), QuoteNode.(args), QuoteNode.(sort(args))
+
+    @assert isdefined(__module__, f_sym)
+
+    f = getproperty(__module__, f_sym)
```
Comment (mileslucas) [inline src/KeywordCalls.jl:47]: should these be `f_sym` or should they be `f_esc`?

**Review by mileslucas — APPROVED** [review]
Looks good to me!

File: src/KeywordCalls.jl (hunk 1 above)
Comment (cscherrer) [reply src/KeywordCalls.jl:47]: IIUC `esc` only makes sense in a quote that's returned by a macro. These lines are not in a quote, so... I think it's ok?

**Pull request merged**